File: files/2024-12-15/CSEC 15.12.R
library(tidyverse)
library(readxl)
path = "files/Excel Challenge 15th Dec.xlsx"
input = read_excel(path, range = "B2:D8")
test  = read_excel(path, range = "E2:E8")

result = input %>%
  mutate(
    Debt = Expense - `Daily Budget`, 
    Cumulative_Debt = accumulate(Debt, ~ max(.x + .y, 0), .init = 0)[-1]
  )

all.equal(result$Cumulative_Debt, test$`Cumulative Debt`)
#> [1] TRUE

File: files/2024-12-15/CSEC 15.12.py
import pandas as pd
from itertools import accumulate

path = "files/Excel Challenge 15th Dec.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=7)
test = pd.read_excel(path, usecols="E", skiprows=1, nrows=7)

input['Debt'] = input['Expense'] - input['Daily Budget']
input['Cumulative_Debt'] = list(accumulate(input['Debt'], lambda cum, debt: max(cum + debt, 0)))

print(input['Cumulative_Debt'].equals(test['Cumulative Debt'])) # True